### ISSUE TITLE
Invalid argument type if using generic callables

### ISSUE THREAD

**Issue by ricardoboss** [vimeo/psalm#6821]
I would expect the following snippet to run without any issues, yet it does: https://psalm.dev/r/2b1f17cd94

I have tested with `dev-master` as well as with the latest release version.

**Comment by danog** [comment]
Something like this should work OK: https://psalm.dev/r/7c189dd28c

**Comment by ricardoboss** [comment]
Thanks for the workaround, but it feels kinda hacky... Do you agree that it _should_ work with my code? I would also be fine if @danog's version was the intended way to handle this, but I want some kind of confirmation.

**Comment by danog** [comment]
Actually, it's not hacky at all, I specifically implemented the templated `@psalm-this-out` for this exact usecase.
The issues psalm is raising with your example are not a bug, it's intended behavior: without `@psalm-this-out` or at least a `@var` declaration, psalm has no way of knowing what generic type do you want `$arr` to contain.

`@psalm-self-out` is a much safer, easier and portable way of dealing with this pattern.

**Comment by ricardoboss** [comment]
I would argue that Psalm is able to infer the generic parameter `T` for `ArrayList` from the return type of the callback (`TOut`). This new `ArrayList` then allows types of `TOut` in the `add` method.

Do you agree?

**Comment by danog** [comment]
Psalm correctly knows that the return type must be an `ArrayList<TOut>`, but it has no way of verifying that the specific instance of `$arr = new ArrayList` will indeed contain only `TOut` elements.

Psalm doesn't know `T` what will `$arr` contain, unless you let it know via either:
- A `@var ArrayList<TOut>` when instantiating `new ArrayList` (which is hacky and a bit ugly)
- Or `@psalm-this-out`, which tells psalm "look, I'm adding a new type to this ArrayList, please expand it to include the passed type".

**Comment by ricardoboss** [comment]
Ah, we might have a misunderstanding here: the intention is to create a new instance of `ArrayList`, which indeed has the explicit type `ArrayList<TOut>` in this case. I don't want to expand the generic type or add an element with a new type.

I want to create a list whose generic parameter is inferred from a callback. Psalm must then verify that the callback only returns instances of `TOut` which in turn verifies only `TOut` instances are added to the new list.

**Comment by ricardoboss** [comment]
I also think this should defintely work when given the type explicitly: https://psalm.dev/r/640f73d840

Psalm still uses the wrong template parameter type.

**Comment by ricardoboss** [comment]
Any updates on this?

**Comment by orklah** [comment]
In your example, there is no warranty that TOut is compatible with T, so Psalm is right not to allow this. If we change this we get: https://psalm.dev/r/f2c3e0e105

Because Psalm complains it doesn't know the type of ArrayList. It's a foggy area of Psalm where if you don't initialize your template at the creation, it's mixed and it will just accept anything. If we force the initialization, there is no more issues:
https://psalm.dev/r/c8bda24b06

Does this solve your issue?

**Comment by ricardoboss** [comment]
Okay, but why would TOut need to be compatible with T?

Say I have an `ArrayList<int> $a` and will now call it with this callback: `$b = $a->map(fn($v) => (string)$v);`.
`$b` should now only contain strings, right? So it's type is `ArrayList<string>` (or even `ArrayList<non-empty-string>`).

In no way must `T` be compatible to `TOut`.

**Comment by orklah** [comment]
Well, given that TOut is pushed to ArrayList::add that expects a T, it's coherent to have TOut to be a T.

Maybe it's more clear if you add `of int` to your `@template T`: suddenly, you're forced to have a TOut that is an int to be accepted by  ArrayList::add.

The issue is you have three template (T from your current class, TOut from your callable, T from the ArrayList you're creating) and two are resolved the same way, that makes TOut necessarily a subset of T

**Comment by ricardoboss** [comment]
So, correct me if I'm wrong, but template parameters have a scope, right? And although `T` is given from the scope where `map` is called, the `T` for the new `ArrayList` is resolved WITHIN the map function. Right?

So even though the generic parameters have the same _name_, their value is resolved _per scope_, right?

Let me reiterate: the `ArrayList<T>` is able to have a completely unrelated template parameter value than the `ArrayList<T>` on which `map` is invoked.

You are telling me the same thing again and again and I'm not sure we mean the same thing. I think I understand what you mean, but that is not what I want to achieve.

**Comment by ricardoboss** [comment]
Look at this again: https://github.com/vimeo/psalm/issues/6821#issuecomment-975528863

The `ArrayList<TOut>` is what I actually want. Calling `add` should accept a `TOut` variable, correct? Then why is Psalm complaining it should be of type `T`?

**Comment by ricardoboss** [comment]
To be clear: `TOut` is in no way meant to be related to `T`. It should not be a covariant to `T`. This is just my name for it. It could aswell be called `U` or any other letter.

**Comment by orklah** [comment]
Psalm seems to resolve both T (from current class and from new instance) the same way. For example, https://psalm.dev/r/677a7a05d1 does not produce issue because T from ArrayList2 is treated as totally unrelated to T from ArrayList

**Comment by ricardoboss** [comment]
Exactly. Would you agree that this behavior _should_ be different? Can I help in any way?

**Comment by orklah** [comment]
I'm not sure if there is a reason why it's like that. I wouldn't mind having it changed if our CI pass, but I'm not actually sure where to look unfortunately

**Comment by ricardoboss** [comment]
I could add a test for this specific behavior. Then the CI would fail even though this is a perfectly valid use-case 🙃

**Comment by orklah** [comment]
Well, I don't have the knowledge on this and currently don't have the time to go on a bug hunt on a topic this specific. It will have to wait until someone cares enough to try and fix it...

**Comment by ricardoboss** [comment]
Understandable. I want to try fixing this, but I suppose this isn't an easy task. Where can I ask if I need help?

**Comment by orklah** [comment]
Please ask here, if you can find something weird or at least the code that handle this, we can try to help understand or give our input
